### PR TITLE
Callbacks for actions and connection

### PR DESCRIPTION
**Pull request by alecdotninja**
This PR contains one way to address #20 . It is draft because I want to make sure that this is a feature that we want to add before I take the time to update the test suite.

It introduces the new idea of an "action" (basically anything that Motion does to a component while it is connected) and adds callbacks for it (`before_action`, `around_action`, and `after_action`). For consistency, the existing lifecycle methods for connection have also been lifted into callbacks (`after_connect` and `after_disconnect`).

#### Pros
* This would facilitate easier usage of APIs that depend on thread globals in components like [`Time.use_zone`](https://api.rubyonrails.org/classes/Time.html#method-c-use_zone), [`I18n.with_locale`](https://guides.rubyonrails.org/i18n.html#managing-the-locale-across-requests), [`Current.user`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html), [`Papertrail.request.whodunnit`](https://github.com/paper-trail-gem/paper_trail#4-saving-more-information-about-versions) and friends, etc.
* This would bring Motion into alignment with prior art like [Stimulus Reflex](https://docs.stimulusreflex.com/lifecycle#server-side-reflex-callbacks) which added reflex callbacks in response to [a very similar issue](https://github.com/unabridged/motion/issues/20#issuecomment-653370129).
* This would allow users to DRY-up repeated code at the beginning and/or end of all "actions" on their components (a use-case unrelated to thread global management is not currently known to me, but seems likely to exist).
* This would allow users to avoid [awkward overrides](https://github.com/unabridged/motion/issues/20#issuecomment-653377971) to run code around rendering.

#### Cons
* ~~This feature may only be used by a minority of users but will [add overhead for all](https://github.com/github/view_component/pull/192#issuecomment-580746645).~~ [see [below](https://github.com/unabridged/motion/pull/27#issuecomment-656268176)]
* Callbacks have reputation for being [an anti-pattern in a lot of cases](https://samuelmullen.com/2013/05/the-problem-with-rails-callbacks/). I think the thread global management case is legitimate, but there might exist a better API for that.
* The concept of "action" is vague and was created specifically to address #20. It's possible a better abstraction exists that would handle other use-cases, map more cleanly onto existing ideas in Motion, and be more intuitive for users.
* With 1.0 on the horizon, I'm reluctant to commit to any new features in the public API since it would have to be supported until 2.0 (which I would personally prefer to never happen).

My intention is that this PR can serve as a place for a public discussion about whether adding these (or any) callbacks to Motion is a good idea.

### CODE REVIEW

**Comment by latortuga** [comment]
The only real objection I have is performance related. I'm compelled by "it's useful in some circumstances", it does fit with the Rails way, and we've had a specific request for it. Can you make them optional so we don't incur the performance penalty? Something like:

```ruby
config.callbacks = false
```

And in case that is set to true, require an additional file that overwrites the right methods with a callback-enabled version?

**Comment by Gmfholley** [comment]
I like the naming of this feature.  The code looks good.

I know a lot of other libraries implement features like these.  But I'm unsure if this is a good idea.

I'm concerned about the performance hit.  A configuration is a good idea, but it also adds complexity.

In addition to performance, I am concerned about the maintainability of the gem and the core of what it does.  Simple in a library is good.  It's hard to remove features once they are released.  It adds complexity for users as well.  

Those concerns should be weighed against what users need.  Will people not use this gem because it doesn't have this feature?  In the specific situation where this feature was requested, the user time zone can be handled in other ways that don't require a change to motion.

So that's why I feel unsure about whether this is a good idea.

**Comment by gcentauri** [comment]
I really don't have much context besides following the issue and PR convos, so I don't know the technical details here... but, would it be possible to create this functionality with an "extension" gem? I know there's some libraries out there that are sort of "modular" in this way where additional functionality can exist in its own companion library. This might help satisfy the users who need it while minimizing the need to maintain within the main library and incurring a performance hit. just throwing it out there, i'm hoping to spend more time actually reading the code when i can.

**Comment by alecdotninja** [comment]
> Can you make them optional so we don't incur the performance penalty?

Reading the config probably would not work well for this because we either have to delay until runtime (and read it every time or introduce some kind of new state) or read it on class load (which may be too soon for the config to be loaded depending on how the user writes their initializers).

It wouldn't be too technically difficult to make them a separate module that you include instead of or in addition to `Motion::Component`, but (like Wendy said) that is added complexity. If we add this feature, I would rather just _always_ have it.

> I am concerned about the maintainability of the gem and the core of what it does. Simple in a library is good. It's hard to remove features once they are released. It adds complexity for users as well.

Retweet. Most of my concern here is around having this feature at all (Is it worth the complexity?) and the name "actions" specifically (What does it mean? Do we want it to mean this?). I'm also worried that we will regret having this feature in the future.

> [W]ould it be possible to create this functionality with an "extension" gem?

I think this functionality is intimately connected with the core enough that splitting it out into its own gem would be prohibitively costly from a maintenance perspective.

---

Since performance seems to be one of the main concerns, I decided to run [my own benchmark](https://gist.github.com/alecdotninja/95e289d68483534c3a641293bb91c117):

```
                                       user     system      total        real
process motion with callbacks      0.890577   0.000000   0.890577 (  0.889007)
process motion without callbacks   0.911804   0.000000   0.911804 (  0.911910)
```

```
                                       user     system      total        real
process motion with callbacks      0.891913   0.000000   0.891913 (  0.890510)
process motion without callbacks   0.888749   0.000000   0.888749 (  0.888766)
```

```
                                       user     system      total        real
process motion with callbacks      0.912935   0.000000   0.912935 (  0.911904)
process motion without callbacks   0.913645   0.000000   0.913645 (  0.913629)
```

```
                                       user     system      total        real
process motion with callbacks      0.919879   0.000000   0.919879 (  0.918517)
process motion without callbacks   0.921038   0.000000   0.921038 (  0.921055)
```

**tl;dr** On my machine, there is _no_ reliably measurable difference between having callbacks and not for users who do not use them.

**Comment by alecdotninja** [comment]
Thank you for your feedback everyone!

I still have some reservations about whether "action" callbacks are the right choice here, but since no better option has come to mind over the past couple days, the work is mostly done, and there is no major performance impact, I think it makes sense to move forward.

File: lib/motion/component/lifecycle.rb [hunk 1]
```diff
@@ -19,12 +30,85 @@ def upgrade_from(previous_revision, _instance)
...
-      def connected
+      def process_disconnect
+        _run_disconnect_callbacks
+
+        # TODO: Remove at next minor release
+        if respond_to?(:disconnected)
+          ActiveSupport::Deprecation.warn(
+            "The `disconnected` lifecycle method is being replaced by the " \
+            "`after_disconnect` callback and will no longer be automatically " \
+            "invoked in the next **minor release** of Motion."
```
Comment (Gmfholley) [inline lib/motion/component/lifecycle.rb:98]: Thanks for this.
